### ISSUE TITLE
[Serve] `request_timeout_s` is ignored in Serve config

### ISSUE THREAD

**Issue by shrekris-anyscale** [ray-project/ray#37882]
### What happened + What you expected to happen

Ray Serve users can set an end-to-end timeout for requests using the `request_timeout_s` field in the Serve config. However, this setting is currently ignored when users deploy using the config.

The setting should be respected.

### Versions / Dependencies

Ray on the latest `master`.

### Reproduction script

```python
# timeout.py

import time
import logging
from ray import serve

logger = logging.getLogger("ray.serve")

@serve.deployment
def sleep():
    for i in range(20):
        time.sleep(1)
        logger.info(f"{i}")

app = sleep.bind()
```

```yaml
# config.yaml

proxy_location: EveryNode

http_options:

  host: 0.0.0.0

  port: 8000

  request_timeout_s: 1

applications:

- name: app1

  route_prefix: /

  import_path: timeout:app

  runtime_env: {}

  deployments:

  - name: sleep
```

The `sleep` deployment's log after 1 request:

```console
INFO 2023-07-27 17:24:42,417 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 replica.py:664 - Started executing request uiLOzcoxCc
INFO 2023-07-27 17:24:43,422 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 0
INFO 2023-07-27 17:24:44,428 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 1
INFO 2023-07-27 17:24:45,430 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 2
INFO 2023-07-27 17:24:46,433 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 3
INFO 2023-07-27 17:24:47,439 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 4
INFO 2023-07-27 17:24:48,441 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 5
INFO 2023-07-27 17:24:49,444 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 6
INFO 2023-07-27 17:24:50,445 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 7
INFO 2023-07-27 17:24:51,450 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 8
INFO 2023-07-27 17:24:52,454 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 9
INFO 2023-07-27 17:24:53,459 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 10
INFO 2023-07-27 17:24:54,461 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 11
INFO 2023-07-27 17:24:55,467 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 12
INFO 2023-07-27 17:24:56,473 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 13
INFO 2023-07-27 17:24:57,476 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 14
INFO 2023-07-27 17:24:58,481 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 15
INFO 2023-07-27 17:24:59,485 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 16
INFO 2023-07-27 17:25:00,489 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 17
INFO 2023-07-27 17:25:01,495 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 18
INFO 2023-07-27 17:25:02,499 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 timeout.py:11 - 19
INFO 2023-07-27 17:25:02,500 app1_sleep app1_sleep#uIiDYw uiLOzcoxCc / app1 replica.py:691 - __CALL__ OK 20082.9ms
```

The `curl` request hangs for the 20 seconds.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by shrekris-anyscale** [comment]
Re-opening for cherry pick.

**Comment by zhe-thoughts** [comment]
Thanks @shrekris-anyscale @edoakes 

When was this issue introduced? Trying to understand if this is a recent regression, or it's a long existing issue and we need to patch earlier releases too

**Comment by shrekris-anyscale** [comment]
The `request_timeout_s` field [wasn't available in Ray 2.5.1](https://github.com/ray-project/ray/blob/164b25fc9db32b8a13b21730d0e63ec9d6e41fa3/python/ray/serve/schema.py#L488-L514). It was [first introduced in Ray 2.6.0](https://github.com/ray-project/ray/blob/fff074f2f498094ec3cd9c14e8cb87a37d516d95/python/ray/serve/schema.py#L515-L518). This is a recent regression, so we don't need to patch earlier releases.

**Comment by shrekris-anyscale** [comment]
Closing this issue since #37903 was merged.
